### DISCUSSION TITLE
next/dynamic import loads component chunk immediately when using { ssr: true }

### DISCUSSION THREAD

**Discussion by jfarris587** [vercel/next.js#33071] [Help]
Dynamic importing components to reduce the size of the initial bundles, and I'm trying to 'lazy load' the Footer at the bottom of the page. Most people don't scroll down that far, so no need to include the JS in the bundle until a user scrolls down. THEN we can load it.

Yet, we still need content in the footer for SEO purposes.

In order to do this, there is a little bit of custom implementation to have it load only when in view

**Layout.tsx**
```tsx
const Footer = dynamic(() => import('@/components/Footer').then((e) => e.Footer), { ssr: true })

export function Layout({ children }: PropsWithChildren<{}>) {
  return (
    <Fragment>
     <Header />
      {children}
      <Suspense ssr={true}>
        <Footer />
      </Suspense>
    </Fragment>
  )
}
```

**Suspense.tsx**

```tsx
export function Suspense({ children, ssr }: PropsWithChildren<{ ssr: boolean }>) {
  const [inViewRef, isInView] = useInView({ initialInView: false, triggerOnce: true, rootMargin: '100px' })

  return <div ref={inViewRef}>{(isInView || (ssr && isServer())) && children}</div>
}

function isServer() {
  return !(typeof window != 'undefined' && window.document)
}

```

The above works and renders the Footer for SSR... **However, the chunk is immediately requested**

see the second JS file requested:
<img width="319" alt="Screen Shot 2022-01-06 at 11 54 48 AM" src="https://user-images.githubusercontent.com/31263423/148419988-4f35be08-90ee-434e-ad7f-cfb1f0234212.png">

**The above implementation works, if you do not want/need `ssr` rendering. The chunk will then load appropriately when the component comes into view, no sooner.**

So whats the point of `dynamic/import` components for `ssr` if they the JS is immediately requested with all the other bundles?

**Comment by zyqxd** [comment]
+1, also seeing this behaviour. The documentation [here](https://nextjs.org/docs/advanced-features/dynamic-import) is underwhelming.

My guess is that with `{ ssr: true }`, your getting a smaller chunk for loading in parallel, but it behaves as if it's in the main code path.

`{ ssr: false }` is the behaviour that I'd expect from `next/dynamic` where chunks are requested as they're inserted into the dom

**Comment by icyJoseph** [comment]
I think there's a fundamental misunderstanding of what's going on here.

First of all, dynamic (from next) and lazy (from React), under the hood are doing link prefetch and preload.

Here's a verbose break down of these two, from Webpack's perspective: https://webpack.js.org/guides/code-splitting/#prefetchingpreloading-modules.

What you are describing is the ability, to lazy load these scripts, when the viewport is nearby, like it's done with images for example. That's out of the responsibility boundary of dynamic imports. 

The idea of the lazy loading technique is to break apart a large bundle, into chunks that are loaded as needed by the document. The fact that the main bundle is broken into chunks, means less joint time is spent parsing and rendering the first page, and gives room for interactivity. When the chunks are loaded these are again parsed and rendering into the browser is done, at a later time.

Chunking is not as simple as breaking into dynamic modules, sometimes it can hurt more than it helps. Dynamic modules have overhead, and in total, more time is spend parsing a main bundle with a bunch of chunks, compared to a single monolith bundle, the advantage is that there's breathing room in between the main bundle and chunk loading (this is a fair assumption).

You can do this by implementing two event listeners for scroll and resize, and calculate whether or not a placeholder is on the viewport as the events fire, as soon as you come near the viewport, let react know to render the dynamic/lazy component, only then the chunk would be required. The https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API, also helps out on this regard.

**Reply by icyJoseph** [reply]
Here's a dummy example of what I meant on the last paragraph:

```tsx
import { RefObject, useEffect, useState, ReactElement, useRef } from "react";

import type { NextPage } from "next";
import dynamic from "next/dynamic";
import Link from "next/link";

const LazyButton = dynamic(() => import("../components/Button"));

interface Args extends IntersectionObserverInit {
  freezeOnceVisible?: boolean;
}

function useIntersectionObserver(
  elementRef: RefObject<Element>,
  {
    threshold = 0,
    root = null,
    rootMargin = "0%",
    freezeOnceVisible = false,
  }: Args
): IntersectionObserverEntry | undefined {
  const [entry, setEntry] = useState<IntersectionObserverEntry>();

  const frozen = entry?.isIntersecting && freezeOnceVisible;

  const updateEntry = ([entry]: IntersectionObserverEntry[]): void => {
    setEntry(entry);
  };

  const stableThreshold = Array.isArray(threshold)
    ? threshold.join(";")
    : threshold;

  useEffect(() => {
    const node = elementRef?.current;
    const hasIOSupport = !!window.IntersectionObserver;

    if (!hasIOSupport || frozen || !node) return;

    const inputThreshold =
      typeof stableThreshold === "string"
        ? stableThreshold.split(";").map(Number)
        : stableThreshold;

    const observerParams = { threshold: inputThreshold, root, rootMargin };
    const observer = new IntersectionObserver(updateEntry, observerParams);

    observer.observe(node);

    return () => observer.disconnect();
  }, [elementRef, stableThreshold, root, rootMargin, frozen]);

  return entry;
}

const RenderWithWindowIntersection = ({
  children,
}: {
  children: ({ visible }: { visible: boolean }) => ReactElement;
}) => {
  const ref = useRef<HTMLDivElement | null>(null);
  const entry = useIntersectionObserver(ref, {});
  const visible = Boolean(entry?.isIntersecting);

  return (
    <div ref={ref}>
      <>{children({ visible })}</>
    </div>
  );
};

const Home: NextPage = () => {
  return (
    <div>
      <h1 id="home">Home</h1>

      <Link href="/blog">blog</Link>

      <div style={{ height: "200vh" }}></div>

      <RenderWithWindowIntersection>
        {({ visible }) => {
          return <>{visible ? <LazyButton /> : null}</>;
        }}
      </RenderWithWindowIntersection>
    </div>
  );
};

export default Home;
```

The Button chunk is only loaded when you scroll down to the placeholder.

**Comment by MurkyMeow** [comment]
Because Next.js is doing source code transformation during build-time you really have to be careful about how you write and structure the code.

My guess is that when a dynamic component appears directly (unconditionally) in the JSX the bundle is loaded immediately irrespective of the runtime logic.

I've made an example of getting the code structure right so it lazy-loads properly: https://github.com/Fractal-Web/next-ssr-hydration
